File: apprain_barcode_scanner_android/android/src/main/kotlin/com/apprain/barcode_scanner/ApprainBarcodeScannerPlugin.kt
package com.apprain.barcode_scanner

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * ApprainBarcodeScannerPlugin — Flutter plugin entry point for Android.
 *
 * Registers the PlatformView factory and sets up MethodChannel / EventChannel
 * communication between Flutter and native CameraX + ML Kit scanning.
 */
class ApprainBarcodeScannerPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var scanEventChannel: EventChannel
    private lateinit var cameraEventChannel: EventChannel
    private var scannerViewFactory: ScannerPlatformViewFactory? = null
    private var activityBinding: ActivityPluginBinding? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "com.apprain.barcode_scanner/methods")
        methodChannel.setMethodCallHandler(this)

        scanEventChannel = EventChannel(binding.binaryMessenger, "com.apprain.barcode_scanner/scan_results")
        cameraEventChannel = EventChannel(binding.binaryMessenger, "com.apprain.barcode_scanner/camera_events")

        scannerViewFactory = ScannerPlatformViewFactory(
            messenger = binding.binaryMessenger,
            scanEventChannel = scanEventChannel,
            cameraEventChannel = cameraEventChannel,
        )

        binding.platformViewRegistry.registerViewFactory(
            "com.apprain.barcode_scanner/scanner_view",
            scannerViewFactory!!,
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        scannerViewFactory?.dispose()
        scannerViewFactory = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val view = scannerViewFactory?.activeView
        when (call.method) {
            "initialize" -> {
                view?.initialize(call.arguments as? Map<*, *>)
                result.success(null)
            }
            "startCamera" -> {
                view?.startCamera()
                result.success(null)
            }
            "stopCamera" -> {
                view?.stopCamera()
                result.success(null)
            }
            "dispose" -> {
                view?.disposeResources()
                result.success(null)
            }
            "setTorch" -> {
                val enabled = (call.arguments as? Map<*, *>)?.get("enabled") as? Boolean ?: false
                view?.setTorch(enabled)
                result.success(null)
            }
            "setZoom" -> {
                val zoom = ((call.arguments as? Map<*, *>)?.get("zoom") as? Number)?.toFloat() ?: 0f
                view?.setZoom(zoom)
                result.success(null)
            }
            "setFocusPoint" -> {
                val args = call.arguments as? Map<*, *>
                val x = (args?.get("x") as? Number)?.toFloat() ?: 0.5f
                val y = (args?.get("y") as? Number)?.toFloat() ?: 0.5f
                view?.setFocusPoint(x, y)
                result.success(null)
            }
            "switchCamera" -> {
                view?.switchCamera()
                result.success(null)
            }
            "updateConfig" -> {
                view?.updateConfig(call.arguments as? Map<*, *>)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // ── ActivityAware ────────────────────────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        scannerViewFactory?.activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        scannerViewFactory?.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityBinding = binding
        scannerViewFactory?.activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activityBinding = null
        scannerViewFactory?.activity = null
    }
}

File: apprain_barcode_scanner_ios/ios/Classes/ApprainBarcodeScannerPlugin.swift
import Flutter
import UIKit

/// Entry point for the Apprain Barcode Scanner iOS plugin.
///
/// Registers the platform view factory and sets up MethodChannel / EventChannel
/// communication between Flutter and native AVFoundation + Vision scanning.
public class ApprainBarcodeScannerPlugin: NSObject, FlutterPlugin {

    private var scannerViewFactory: ScannerPlatformViewFactory?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ApprainBarcodeScannerPlugin()

        // Method channel for control commands
        let methodChannel = FlutterMethodChannel(
            name: "com.apprain.barcode_scanner/methods",
            binaryMessenger: registrar.messenger()
        )

        // Event channels for scan results and camera events
        let scanEventChannel = FlutterEventChannel(
            name: "com.apprain.barcode_scanner/scan_results",
            binaryMessenger: registrar.messenger()
        )
        let cameraEventChannel = FlutterEventChannel(
            name: "com.apprain.barcode_scanner/camera_events",
            binaryMessenger: registrar.messenger()
        )

        let factory = ScannerPlatformViewFactory(
            messenger: registrar.messenger(),
            scanEventChannel: scanEventChannel,
            cameraEventChannel: cameraEventChannel
        )
        instance.scannerViewFactory = factory

        registrar.register(factory, withId: "com.apprain.barcode_scanner/scanner_view")
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let view = scannerViewFactory?.activeView else {
            result(FlutterError(
                code: "NO_VIEW",
                message: "Camera is not available — scanner view not created yet",
                details: "Method: \(call.method)"
            ))
            return
        }

        switch call.method {
        case "initialize":
            let config = call.arguments as? [String: Any]
            view.initialize(config: config)
            result(nil)

        case "startCamera":
            view.startCamera()
            result(nil)

        case "stopCamera":
            view.stopCamera()
            result(nil)

        case "dispose":
            view.disposeResources()
            result(nil)

        case "setTorch":
            let args = call.arguments as? [String: Any]
            let enabled = args?["enabled"] as? Bool ?? false
            view.setTorch(enabled: enabled)
            result(nil)

        case "setZoom":
            let args = call.arguments as? [String: Any]
            let zoom = (args?["zoom"] as? NSNumber)?.floatValue ?? 0.0
            view.setZoom(CGFloat(zoom))
            result(nil)

        case "setFocusPoint":
            let args = call.arguments as? [String: Any]
            let x = (args?["x"] as? NSNumber)?.floatValue ?? 0.5
            let y = (args?["y"] as? NSNumber)?.floatValue ?? 0.5
            view.setFocusPoint(x: CGFloat(x), y: CGFloat(y))
            result(nil)

        case "switchCamera":
            view.switchCamera()
            result(nil)

        case "updateConfig":
            let config = call.arguments as? [String: Any]
            view.updateConfig(config: config)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
